File: Week_03/47.permutations-ii.go
/*
 * @lc app=leetcode.cn id=47 lang=golang
 *
 * [47] 全排列 II
 *
 * https://leetcode-cn.com/problems/permutations-ii/description/
 *
 * algorithms
 * Medium (61.93%)
 * Total Accepted:    111.4K
 * Total Submissions: 179.8K
 * Testcase Example:  '[1,1,2]'
 *
 * 给定一个可包含重复数字的序列，返回所有不重复的全排列。
 * 
 * 示例:
 * 
 * 输入: [1,1,2]
 * 输出:
 * [
 * ⁠ [1,1,2],
 * ⁠ [1,2,1],
 * ⁠ [2,1,1]
 * ]
 * 
 */
func permuteUnique(nums []int) [][]int {
    sort.Ints(nums)
    result := [][]int{}
    visited := make([]int, len(nums), len(nums))
    item := []int{}
    var build func(first int)
    build = func(first int) {
        if first == len(nums) {
            result = append(result, append([]int{}, item...))
            return
        }
        for i := 0; i < len(nums); i++ {
            if visited[i] == 1 || (i > 0 && nums[i - 1] == nums[i] && visited[i - 1] == 1) {
                continue
            }
            item = append(item, nums[i])
            visited[i] = 1
            build(first + 1)
            visited[i] = 0
            item = item[:len(item) - 1]
        }
    }
    build(0)
    return result
}

File: Week_03/47.permutations-ii.java
/*
 * @lc app=leetcode.cn id=47 lang=java
 *
 * [47] 全排列 II
 *
 * https://leetcode-cn.com/problems/permutations-ii/description/
 *
 * algorithms
 * Medium (61.93%)
 * Total Accepted:    111.4K
 * Total Submissions: 179.8K
 * Testcase Example:  '[1,1,2]'
 *
 * 给定一个可包含重复数字的序列，返回所有不重复的全排列。
 * 
 * 示例:
 * 
 * 输入: [1,1,2]
 * 输出:
 * [
 * ⁠ [1,1,2],
 * ⁠ [1,2,1],
 * ⁠ [2,1,1]
 * ]
 * 
 */
class Solution {
    private List<List<Integer>> result = new ArrayList<>();
    private boolean[] visited;
    private List<Integer> item = new ArrayList<>();

    private void permute(int[] nums, int first) {
        if (first == nums.length) {
            result.add(new ArrayList<Integer>(item));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (visited[i] || (i > 0 && nums[i] == nums[i-1] && visited[i-1])) {
                continue;
            }
            visited[i] = true;
            item.add(nums[i]);
            permute(nums, first + 1);
            item.remove(item.size() - 1);
            visited[i] = false;
        }
    }

    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        visited = new boolean[nums.length];
        Arrays.fill(visited, false);
        permute(nums, 0);
        return result;
    }
}
